pkg/operator/kubelet.go

```diff
@@ -0,0 +1,151 @@
+package operator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/grafana/agent/pkg/operator/clientutil"
+	"github.com/grafana/agent/pkg/operator/logutil"
+	core_v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	controller "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type kubeletReconciler struct {
+	client.Client
+	kubeletNamespace, kubeletName string
+}
+
+func (r *kubeletReconciler) Reconcile(ctx context.Context, req controller.Request) (res controller.Result, err error) {
+	l := logutil.FromContext(ctx)
+	level.Info(l).Log("msg", "reconciling node")
+
+	var nodes core_v1.NodeList
+	if err := r.List(ctx, &nodes); err != nil {
+		level.Error(l).Log("msg", "failed to list nodes for kubelet service", "err", err)
+		return res, fmt.Errorf("unable to list nodes: %w", err)
+	}
+	nodeAddrs, err := getNodeAddrs(l, &nodes)
+	if err != nil {
+		level.Error(l).Log("msg", "could not get addresses from all nodes", "err", err)
+		return res, fmt.Errorf("unable to get addresses from nodes: %w", err)
+	}
+
+	labels := mergeMaps(managedByOperatorLabels, map[string]string{
+		// Labels taken from prometheus-operator:
+		// https://github.com/prometheus-operator/prometheus-operator/blob/2c81b0cf6a5673e08057499a08ddce396b19dda4/pkg/prometheus/operator.go#L586-L587
+		"k8s-app":                "kubelet",
+		"app.kubernetes.io/name": "kubelet",
+	})
+
+	svc := &core_v1.Service{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      r.kubeletName,
+			Namespace: r.kubeletNamespace,
+			Labels:    labels,
+		},
+		Spec: core_v1.ServiceSpec{
+			Type:      core_v1.ServiceTypeClusterIP,
+			ClusterIP: "None",
+			Ports: []core_v1.ServicePort{
+				{Name: "https-metrics", Port: 10250},
+				{Name: "http-metrics", Port: 10255},
+				{Name: "cadvisor", Port: 4194},
+			},
+		},
+	}
+
+	eps := &core_v1.Endpoints{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      r.kubeletName,
+			Namespace: r.kubeletNamespace,
+			Labels:    labels,
+		},
+		Subsets: []core_v1.EndpointSubset{{
+			Addresses: nodeAddrs,
+			Ports: []core_v1.EndpointPort{
+				// Taken from https://github.com/prometheus-operator/prometheus-operator/blob/2c81b0cf6a5673e08057499a08ddce396b19dda4/pkg/prometheus/operator.go#L593
+				{Name: "https-metrics", Port: 10250},
```

Comment:
Are these ports guaranteed?

Reply:
I don't think so. They're the defaults but you can technically change each Kubelet to have a different one. I'm not sure there's a good way to support that, and I also can't find a way to access it from within the cluster.

Reply:
Seems good enough and we can always see what responses we get